plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,472 @@
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_ccgx_pure_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	if (!fu_ccgx_pure_hid_magic_unlock(self, error))
+		return FALSE;
+	if (!fu_ccgx_pure_hid_ensure_fw_info(self, error))
+		return FALSE;
+
+	fu_device_add_instance_strup(device,
+				     "MODE",
+				     fu_ccgx_pure_hid_fw_mode_to_string(self->operating_mode));
+	if (!fu_device_build_instance_id(FU_DEVICE(self), error, "USB", "VID", "PID", "MODE", NULL))
+		return FALSE;
+
+	fu_device_add_instance_u16(FU_DEVICE(self), "SID", self->silicon_id);
+	if (!fu_device_build_instance_id_full(FU_DEVICE(self),
+					      FU_DEVICE_INSTANCE_FLAG_QUIRKS,
+					      error,
+					      "CCGX",
+					      "SID",
+					      NULL))
+		return FALSE;
+	g_debug("got silicon ID: 0x%04x", self->silicon_id);
+
+	/* ensure the remove delay is set, even if no quirk matched */
+	if (fu_device_get_remove_delay(FU_DEVICE(self)) == 0)
+		fu_device_set_remove_delay(FU_DEVICE(self), 5000);
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_ccgx_pure_hid_device_prepare_firmware(FuDevice *device,
+					 GInputStream *stream,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuCcgxPureHidDevice *self = FU_CCGX_PURE_HID_DEVICE(device);
+	FuCcgxFwMode fw_mode;
+	GPtrArray *records = NULL;
+	FuCcgxFirmwareRecord *vidpid_rcd = NULL;
+	guint16 fw_silicon_id;
+	gsize fw_size = 0;
+	guint16 vid;
+	guint16 pid;
+	const guint8 *bytes;
+	g_autoptr(FuFirmware) firmware = fu_ccgx_firmware_new();
+
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+
+	/* check the silicon ID */
+	fw_silicon_id = fu_ccgx_firmware_get_silicon_id(FU_CCGX_FIRMWARE(firmware));
+	if (fw_silicon_id != self->silicon_id) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "silicon id mismatch, expected 0x%x, got 0x%x",
+			    self->silicon_id,
+			    fw_silicon_id);
+		return NULL;
+	}
+
+	fw_mode = fu_ccgx_firmware_get_fw_mode(FU_CCGX_FIRMWARE(firmware));
+	if (fw_mode != fu_ccgx_fw_mode_get_alternate((FuCcgxFwMode)self->operating_mode)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "FuCcgxFwMode mismatch, expected %s, got %s",
+			    fu_ccgx_fw_mode_to_string(
+				fu_ccgx_fw_mode_get_alternate((FuCcgxFwMode)self->operating_mode)),
+			    fu_ccgx_fw_mode_to_string(fw_mode));
+		return NULL;
+	}
+
+	/* validate all records has proper size */
+	records = fu_ccgx_firmware_get_records(FU_CCGX_FIRMWARE(firmware));
+	g_debug("records found: %u", records->len);
+
+	for (guint i = 0; i < records->len; i++) {
+		FuCcgxFirmwareRecord *record = g_ptr_array_index(records, i);
+		gsize record_size = g_bytes_get_size(record->data);
+		if (record_size != self->flash_row_size) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "expected block length %lu, got %lu: array id=0x%02x, "
+				    "row=0x%04x (:%02x%04x%04lx)",
```

Comment:
this isn't the correct format string for architecture independent, look at the CI error.

Reply:
ya, just coerce into (guint) to print %x